From 36afdcc2a32e55d82fde876058c3ac948f80c32c Mon Sep 17 00:00:00 2001
From: "Signed-off-by: Fancy Fang" <chen.fang@freescale.com>
Date: Tue, 1 Nov 2016 16:33:19 +0200
Subject: [PATCH 0269/5242] MLK-11395-4 ARM: imx: add mipi phy regulator
 handling to gpcv2

commit  9764c0ca5d64543cbd06424026f14015aceefb09 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add mipi phy regulator notify callback to power on or power off this
phy along with the regulator enable/disable called. This will be used
by mipi dsi/csi later.

Signed-off-by: Fancy Fang <chen.fang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/boot/dts/imx7d.dtsi |    1 +
 arch/arm/mach-imx/gpcv2.c    |   52 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 53 insertions(+)

diff --git a/arch/arm/boot/dts/imx7d.dtsi b/arch/arm/boot/dts/imx7d.dtsi
index 675186d..f362386 100644
--- a/arch/arm/boot/dts/imx7d.dtsi
+++ b/arch/arm/boot/dts/imx7d.dtsi
@@ -169,6 +169,7 @@
 		#interrupt-cells = <3>;
 		interrupt-parent = <&intc>;
 		fsl,mf-mix-wakeup-irq = <0x54010000 0xc00 0x0 0x1040640>;
+		mipi-phy-supply = <&reg_1p0d>;
 	};
 };
 
diff --git a/arch/arm/mach-imx/gpcv2.c b/arch/arm/mach-imx/gpcv2.c
index 02a2524..dae97ad 100644
--- a/arch/arm/mach-imx/gpcv2.c
+++ b/arch/arm/mach-imx/gpcv2.c
@@ -108,6 +108,7 @@ enum imx_gpc_slot {
 static u32 gpcv2_mf_irqs[IMR_NUM];
 static u32 gpcv2_mf_request_on[IMR_NUM];
 static DEFINE_SPINLOCK(gpcv2_lock);
+static struct notifier_block nb_mipi;
 
 static int imx_gpcv2_irq_set_wake(struct irq_data *d, unsigned int on)
 {
@@ -582,6 +583,37 @@ static int imx_gpcv2_domain_alloc(struct irq_domain *domain,
 	.free	= irq_domain_free_irqs_common,
 };
 
+static int imx_mipi_regulator_notify(struct notifier_block *nb,
+					unsigned long event,
+					void *ignored)
+{
+	u32 val = 0;
+
+	switch (event) {
+	case REGULATOR_EVENT_PRE_DO_ENABLE:
+		val = readl_relaxed(gpc_base + GPC_PGC_CPU_MAPPING);
+		writel_relaxed(val | BIT(2), gpc_base + GPC_PGC_CPU_MAPPING);
+
+		val = readl_relaxed(gpc_base + GPC_PU_PGC_SW_PUP_REQ);
+		writel_relaxed(val | BIT(0), gpc_base + GPC_PU_PGC_SW_PUP_REQ);
+		break;
+	case REGULATOR_EVENT_PRE_DO_DISABLE:
+		val = readl_relaxed(gpc_base + GPC_PU_PGC_SW_PDN_REQ);
+		writel_relaxed(val | BIT(0), gpc_base + GPC_PU_PGC_SW_PDN_REQ);
+
+		val = readl_relaxed(gpc_base + GPC_PGC_MIPI_PHY);
+		writel_relaxed(val | BIT(0), gpc_base + GPC_PGC_MIPI_PHY);
+
+		val = readl_relaxed(gpc_base + GPC_PGC_CPU_MAPPING);
+		writel_relaxed(val & ~BIT(2), gpc_base + GPC_PGC_CPU_MAPPING);
+		break;
+	default:
+		break;
+	}
+
+	return NOTIFY_OK;
+}
+
 static int __init imx_gpcv2_init(struct device_node *node,
 			       struct device_node *parent)
 {
@@ -692,6 +724,26 @@ void __init imx_gpcv2_check_dt(void)
 
 static int imx_gpcv2_probe(struct platform_device *pdev)
 {
+	int ret;
+	struct regulator *mipi_reg;
+
+	if (cpu_is_imx7d()) {
+		mipi_reg = devm_regulator_get(&pdev->dev, "mipi-phy");
+		if (IS_ERR(mipi_reg)) {
+			ret = PTR_ERR(mipi_reg);
+			dev_info(&pdev->dev, "mipi regulator not ready.\n");
+			return ret;
+		}
+		nb_mipi.notifier_call = &imx_mipi_regulator_notify;
+
+		ret = regulator_register_notifier(mipi_reg, &nb_mipi);
+		if (ret) {
+			dev_err(&pdev->dev,
+				"mipi regulator notifier request failed.\n");
+			return ret;
+		}
+	}
+
 	return 0;
 }
 
-- 
1.7.9.5

